src/mono/wasm/runtime/loader/assets.ts

```diff
@@ -422,9 +552,73 @@ function download_resource(request: ResourceRequest): LoadingResource {
     }
 }
 
+async function download_resource_with_cache(request: ResourceRequest): Promise<Response> {
+    let response = await findCachedResponse(request);
+    if (!response) {
+        response = await fetchResource(request);
+        addCachedReponse(request, response);
+    }
+
+    return response;
+}
+
+const monoToBlazorAssetTypeMap: { [key: string]: WebAssemblyBootResourceType | undefined } = {
+    "resource": "assembly",
+    "assembly": "assembly",
+    "pdb": "pdb",
+    "icu": "globalization",
+    "vfs": "configuration",
+    "dotnetwasm": "dotnetwasm",
+};
+
+const credentialsIncludeAssetBehaviors: AssetBehaviors[] = ["vfs"]; // Previously only configuration
+
+function fetchResource(request: ResourceRequest): Promise<Response> {
+    // Allow developers to override how the resource is loaded
+    let url = request.resolvedUrl!;
+    if (loaderHelpers.loadBootResource) {
+        const resourceType = monoToBlazorAssetTypeMap[request.behavior];
+        if (resourceType) {
+            const customLoadResult = loaderHelpers.loadBootResource(resourceType!, request.name, url, request.hash ?? "");
+            if (customLoadResult instanceof Promise) {
+                // They are supplying an entire custom response, so just use that
+                return customLoadResult;
+            } else if (typeof customLoadResult === "string") {
+                // They are supplying a custom URL, so use that with the default fetch behavior
+                url = customLoadResult;
+            }
+        }
+    }
+
+    // Note that if cacheBootResources was explicitly disabled, we also bypass hash checking
+    // This is to give developers an easy opt-out from the entire caching/validation flow if
+    // there's anything they don't like about it.
+    const fetchOptions: RequestInit = {
+        cache: "no-cache"
+    };
+
+    if (credentialsIncludeAssetBehaviors.includes(request.behavior)) {
+        // Include credentials so the server can allow download / provide user specific file
+        fetchOptions.credentials = "include";
+    } else {
+        // Any other resource than configuration should provide integrity check
+        fetchOptions.integrity = shouldApplyIntegrity() ? (request.hash ?? "") : undefined;
```

Comment:
Why would we skip integrity when cache is off ? Also please rename `shouldApplyIntegrity` so that it mentions cache.

Reply:
```
// Note that if cacheBootResources was explicitly disabled, we also bypass hash checking
// This is to give developers an easy opt-out from the entire caching/validation flow if
// there's anything they don't like about it.
```

Reply:
That sounds like excuse for not having another flag.

Reply:
Maybe 🤷‍♂️ I don't have strong opinion here